### ISSUE TITLE
[Core Request] Cemu - Wii U emulator

### ISSUE THREAD

**Issue by almi05** [libretro/RetroArch#14416]
Now that Cemu's source code has been released and Wii U is discontinued, it would be nice to integrate it into a RetroArch core and be ported to Xbox and mobile devices too!

**Comment by TheRetroRepairGuy** [comment]
Indeed I very much agree. My own computer can't run cemu well so I'm curious as to if I would run well on my series x.

**Comment by ADHDefy** [comment]
I am not sure if this is how this works, but it would also be great to see the option to store unpacked/rpx game folders compressed in a single-file format (CHD? 7z or zip?) and have them readable, making it more convenient to just throw games into the ROM folder while still saving drive space. 

I am not sure if that would be up to the CEMU devs to make possible, but I thought I'd put it out there.

**Comment by ichee** [comment]
Core requests may be better supported in the forums.

Core request topic is at:

https://forums.libretro.com/t/core-and-feature-requests/5895.

**Comment by rosostolato** [comment]
What about now that finally CEMU is open source? https://github.com/cemu-project/Cemu

**Comment by hizzlekizzle** [comment]
No one is working on such a thing currently, as far as I'm aware, but we would certainly welcome it if someone wants to take a stab at it.

**Comment by rosostolato** [comment]
I'm a developer with some c++ background but I'm not sure if I can handle a project so big, I'd like to help but I don't even know where to start.

**Comment by hizzlekizzle** [comment]
yeah, it's going to be a big one, and it will require familiarity with hardware-rendering APIs and threading, at least. Even if that's outside of your skillset, keep an eye out in case someone else gets started on it, as there's always room for contributions :)

**Comment by Fijxu** [comment]
+1

**Comment by rosostolato** [comment]
> hello to give you some hope i am currently working on it i am in the test phase it is very tedious and my time is quite limited

Do you have it on github?

**Comment by bactino** [comment]
not at the moment it is not stable at all too many bugs when I have something stable enough I will post it

**Comment by hizzlekizzle** [comment]
oh wow, that's great news. Thanks for taking a stab at it!

**Comment by bannmannutt** [comment]
So how's that coming along?

**Comment by thanone2** [comment]
> hello to give you some hope i am currently working on it i am in the test phase it is very tedious and my time is quite limited

How far are you on thi? Would love to use this on steam on my Atari VCS.

**Comment by Focus942** [comment]
Please someone make Wii U on RetroArch be Real!

**Comment by LADsy8** [comment]
> not at the moment it is not stable at all too many bugs when I have something stable enough I will post it

how's it going?

**Comment by paulo101977** [comment]
> Please someone make Wii U on RetroArch be Real!

I also need this support to be able to train AI agents.

**Comment by tacotopasu** [comment]
This is so needed! It's been a while since cemu has been open sourced now. Although making it into a core might not be super straight forward, it can be done! It would be such a welcome core.

**Comment by gouchi** [comment]
For reference, upstream [issue](https://github.com/cemu-project/Cemu/issues/1160).

There are currently [two](https://github.com/zorn-v/Cemu-libretro) [attempts](https://github.com/paulo101977/Cemu-libretro).

**Comment by paulo101977** [comment]
> For reference, upstream [issue](https://github.com/cemu-project/Cemu/issues/1160).
> 
> There are currently [two](https://github.com/zorn-v/Cemu-libretro) [attempts](https://github.com/paulo101977/Cemu-libretro).

One of them is mine, but I interrupted it to work on Xemu. However, I also encountered a barrier and had to shift my focus solely to its use with AI. With Cemu, I had to stop because I needed the save state for AI training, and it hasn't been implemented by their team yet, and probably never will be.

**Comment by danprice142** [comment]
Well the wait is over I`ve only hooked up OpenGL and its only been tested on Windows https://github.com/danprice142/Cemu-Libretro @paulo101977 check my github I saw you were looking for a RPCS3 core and if it helps your project Eden just merged a Libretro Core

**Comment by paulo101977** [comment]
> Well the wait is over I`ve only hooked up OpenGL and its only been tested on Windows https://github.com/danprice142/Cemu-Libretro [@paulo101977](https://github.com/paulo101977) check my github I saw you were looking for a RPCS3 core and if it helps your project Eden just merged a Libretro Core

Hey. Thanks. But I need a core that saves states, even to be able to use it with reinforcement training (AI).

**Comment by davidhedlund** [comment]
The wait is over, Cemu is finally in **Core Downloader** (remember to **Update Core Info Files**)!

I just tested it on GNU/Linux: https://buildbot.libretro.com/nightly/linux/x86_64/latest/cemu_libretro.so.zip

**Comment by xangelix** [comment]
Yayy!!! Congrats everyone. 😁🎉🎉

**Comment by paulo101977** [comment]
> The wait is over, Cemu is finally in **Core Downloader** (remember to **Update Core Info Files**)!
> 
> I just tested it on GNU/Linux: https://buildbot.libretro.com/nightly/linux/x86_64/latest/cemu_libretro.so.zip

Do you already have his working repository? I looked for it but couldn't find it (I searched the web). I need to fork it and build upon it for my Reinforcement Learning projects.

**Comment by davidhedlund** [comment]
> > The wait is over, Cemu is finally in **Core Downloader** (remember to **Update Core Info Files**)!
> > I just tested it on GNU/Linux: https://buildbot.libretro.com/nightly/linux/x86_64/latest/cemu_libretro.so.zip
> 
> Do you already have his working repository? I looked for it but couldn't find it (I searched the web). I need to fork it and build upon it for my Reinforcement Learning projects.

https://github.com/libretro/libretro-super/blob/master/dist/info/cemu_libretro.info says `authors = "Cemu Project|WizzardSK"`

So I guess the repo is https://github.com/WizzardSK/cemu-libretro

**Comment by gouchi** [comment]
@almi05 Please close this issue as the core is available on the buildbot albeit it is is wip.

Thank you.

**Comment by paulo101977** [comment]
> > > The wait is over, Cemu is finally in **Core Downloader** (remember to **Update Core Info Files**)!
> > > I just tested it on GNU/Linux: https://buildbot.libretro.com/nightly/linux/x86_64/latest/cemu_libretro.so.zip
> > 
> > 
> > Do you already have his working repository? I looked for it but couldn't find it (I searched the web). I need to fork it and build upon it for my Reinforcement Learning projects.
> 
> https://github.com/libretro/libretro-super/blob/master/dist/info/cemu_libretro.info says `authors = "Cemu Project|WizzardSK"`
> 
> So I guess the repo is https://github.com/WizzardSK/cemu-libretro

Thank u!!!! That’s already a huge step forward. I’m just not sure if it includes features like save states yet, because that would really help me with training AI agents using reinforcement learning.

**Comment by davidhedlund** [comment]
> > > > The wait is over, Cemu is finally in **Core Downloader** (remember to **Update Core Info Files**)!
> > > > I just tested it on GNU/Linux: https://buildbot.libretro.com/nightly/linux/x86_64/latest/cemu_libretro.so.zip
> > > 
> > > 
> > > Do you already have his working repository? I looked for it but couldn't find it (I searched the web). I need to fork it and build upon it for my Reinforcement Learning projects.
> > 
> > 
> > https://github.com/libretro/libretro-super/blob/master/dist/info/cemu_libretro.info says `authors = "Cemu Project|WizzardSK"`
> > So I guess the repo is https://github.com/WizzardSK/cemu-libretro
> 
> Thank u!!!! That’s already a huge step forward. I’m just not sure if it includes features like save states yet, because that would really help me with training AI agents using reinforcement learning.

Well https://github.com/WizzardSK/cemu-libretro/blob/libretro/README.md#known-issues says:

> Save states are not supported - Cemu has no savestate infrastructure (no serialization of PPC/MMU/GPU/HLE state). retro_serialize_size returns 0 by design.

**Comment by paulo101977** [comment]
> > > > > The wait is over, Cemu is finally in **Core Downloader** (remember to **Update Core Info Files**)!
> > > > > I just tested it on GNU/Linux: https://buildbot.libretro.com/nightly/linux/x86_64/latest/cemu_libretro.so.zip
> > > > 
> > > > 
> > > > Do you already have his working repository? I looked for it but couldn't find it (I searched the web). I need to fork it and build upon it for my Reinforcement Learning projects.
> > > 
> > > 
> > > https://github.com/libretro/libretro-super/blob/master/dist/info/cemu_libretro.info says `authors = "Cemu Project|WizzardSK"`
> > > So I guess the repo is https://github.com/WizzardSK/cemu-libretro
> > 
> > 
> > Thank u!!!! That’s already a huge step forward. I’m just not sure if it includes features like save states yet, because that would really help me with training AI agents using reinforcement learning.
> 
> Well https://github.com/WizzardSK/cemu-libretro/blob/libretro/README.md#known-issues says:
> 
> > Save states are not supported - Cemu has no savestate infrastructure (no serialization of PPC/MMU/GPU/HLE state). retro_serialize_size returns 0 by design.

Yes, the Cemu team itself is aware of this issue. Serializing game saves would be a nightmare, as the disk footprint is huge—specifically, serializing memory, states, and so on.
